File: pascal/pascal/ast.go
package pascal

type Node interface {
	node()
}

type Number struct {
	Token Token
	Value string
}

func (n Number) node() {}

type BinOp struct {
	Left  Node
	Op    Token
	Right Node
}

func (b BinOp) node() {}

type UnaryOp struct {
	Op   Token
	Expr Node
}

func (u UnaryOp) node() {}

type Compound struct {
	Children []Node
}

func (c Compound) node() {}

type Assign struct {
	Left  Var
	Op    Token
	Right Node
}

func (a Assign) node() {}

type Var struct {
	Token Token
	Value string
}

func (v Var) node() {}

type NoOp struct{}

func (n NoOp) node() {}

File: interpreter_full/ast.py
from .token import Token

class Node:
    pass

class Number(Node):

    def __init__(self, token: Token):
        self.token = token

    def __str__(self):
        return f"{self.__class__.__name__}({self.token})"
    
class BinOp(Node):

    def __init__(self, left: Node, op: Token, right: Node):
        self.left = left
        self.op = op
        self.right = right

    def __str__(self):
        return f'BinOp{self.op.value}({self.left}, {self.right})'